### PR TITLE
[Attempt 2] Fixed cChunk::m_Entities corruption upon world travel

### LINKED ISSUE #2724: Crash on multiworld travel
I was in the Nether. I executed `/portal world`

```
[70153d24f3fd4a92|15:28:10] Assertion failed: std::find(m_Entities.begin(), m_Entities.end(), a_Entity) == m_Entities.end(), file /home/cuberite/cuberite/src/Chunk.cpp, line 1944
./Cuberite_debug(_Z15PrintStackTracev+0x23)[0xb0c901]
./Cuberite_debug(_ZN6cChunk9AddEntityEP7cEntity+0xce)[0xa0e5f4]
./Cuberite_debug(_ZN6cChunk20MoveEntityToNewChunkEP7cEntity+0xfe)[0xa0af96]
./Cuberite_debug(_ZN6cChunk4TickENSt6chrono8durationIlSt5ratioILl1ELl1000EEEE+0x3e6)[0xa0ac24]
./Cuberite_debug(_ZN9cChunkMap11cChunkLayer4TickENSt6chrono8durationIlSt5ratioILl1ELl1000EEEE+0x86)[0xa23ecc]
./Cuberite_debug(_ZN9cChunkMap4TickENSt6chrono8durationIlSt5ratioILl1ELl1000EEEE+0x58)[0xa2347a]
./Cuberite_debug(_ZN6cWorld4TickENSt6chrono8durationIlSt5ratioILl1ELl1000EEEES4_+0x3f6)[0xad8a12]
./Cuberite_debug(_ZN6cWorld11cTickThread7ExecuteEv+0xad)[0xacf425]
./Cuberite_debug(_ZN9cIsThread9DoExecuteEv+0x2f)[0xb01d8d]
./Cuberite_debug(_ZNKSt7_Mem_fnIM9cIsThreadFvvEEclIIEvEEvPS0_DpOT_+0x65)[0xb0338d]
./Cuberite_debug(_ZNSt12_Bind_simpleIFSt7_Mem_fnIM9cIsThreadFvvEEPS1_EE9_M_invokeIILm0EEEEvSt12_Index_tupleIIXspT_EEE+0x43)[0xb032dd]
./Cuberite_debug(_ZNSt12_Bind_simpleIFSt7_Mem_fnIM9cIsThreadFvvEEPS1_EEclEv+0x1d)[0xb031d3]
./Cuberite_debug(_ZNSt6thread5_ImplISt12_Bind_simpleIFSt7_Mem_fnIM9cIsThreadFvvEEPS3_EEE6_M_runEv+0x1c)[0xb03150]
/usr/lib/x86_64-linux-gnu/libstdc++.so.6(+0xb6970)[0x7f3acdeb1970]
/lib/x86_64-linux-gnu/libpthread.so.0(+0x80a4)[0x7f3ace10e0a4]
/lib/x86_64-linux-gnu/libc.so.6(clone+0x6d)[0x7f3acd62104d]
Cuberite_debug: /home/cuberite/cuberite/src/Chunk.cpp:1944: void cChunk::AddEntity(cEntity*): Assertion `0' failed.

```

### PR DESCRIPTION
**Pull request by LogicParrot**
**Attempt 2, previous attempt: #2940**
Fixes #2724
possibly fixes or completely resolves other world travel issues as well.

### CODE REVIEW

File: src/Chunk.h [hunk 1]
```diff
@@ -260,6 +260,9 @@ class cChunk :
 
 	void AddEntity(cEntity * a_Entity);
 	void RemoveEntity(cEntity * a_Entity);
+	/** RemoveEntity is dangerous if the chunk is inside the tick() method because it invalidates the iterator.
+	This will safely remove an entity. Returns true on success. */
```
Comment (LogicParrot) [inline src/Chunk.h]: returns nothing.

**Comment by tigerw** [comment]
How does this work around the stated diagnosis of cChunk not removing anything at all because it doesn't tick? Iterators shouldn't be invalidated since the Entities container is I believe a list, not vector.

**Comment by LogicParrot** [comment]
@tigerw Because of Diagnoses Continued :tm: 

There are two code paths that lead to teleport (cEntity::doMoveWithWorld).

Code path 1:
 - cChunk::tick() ticks all the entities [here](https://github.com/cuberite/cuberite/pull/2941/files#diff-4b84dbafe20b4c913f541b6369929de5R627), and one of the entities decides to teleport inside its own tick method. In that case, the entity marks itself for removal, and is properly removed inside [this removal if block](https://github.com/cuberite/cuberite/pull/2941/files#diff-4b84dbafe20b4c913f541b6369929de5R638) right after it ticks. This was done properly even before my changes.

Code path 2:
- Some force outside the cChunk::tick, like a plugin, asks for a teleport. In that case:
 A: The chunk may not be ticking at all, and the "removal if block" may never get executed.
 B: Direct removal is fine precisely because the chunk is not in the midst of a tick and there is no risk of iterator corruption.

My changes essentially say:
"If it's code path 2, remove directly. If it's code path 1, schedule for removal as we always have."

**Comment by tigerw** [comment]
I guess that's fine. Probably something in the future along the lines of queueing a task to move the entity would look better and be easier to follow, even if it introduces a one-tick delay.

**Comment by LogicParrot** [comment]
@tigerw It's true that it's a list, but this code causes trouble nevertheless:
` cChunk::tick() -> (*it)->tick() -> cEntity::tick > Tell cChunk to delete (it)`

Is there an easy way to do this? Maybe that would be more readable.
` cChunk::tick() -> (*it)->tick() -> cEntity::tick > Tell cChunk to hop to next iterator and then remove (it)`

**Comment by LogicParrot** [comment]
@Seadragon91 would you mind testing this one? Preferably both netherportal travel and `/portal` travel.

**Comment by Seadragon91** [comment]
This works fine now. Had 5 times the crash described in issue #2839.

**Comment by tigerw** [comment]
So I guess merge, it will hopefully be all refactored away soon whatever the case.

**Pull request merged**